### PR TITLE
Dashboard: Lazy load out of view panels 

### PR DESCRIPTION
**Pull request by ryantxu**
duplicating #15310 with a new branch after a merge error

yet-another-lazy-loading-pull-request!

Following @torkelo idea on #15272, this does not use an intersection observer, but (approximately) calculates visibility from the existing scroll events in DashboardGrid.tsx

This seems to work well!

Open questions:
- Is there a better way to get the height?  
- Is the visibility calculation accurate enough?  missing GRID_CELL_VMARGIN may be an issue for very large dashboards

See also: #7452 #5216 #8365 #8500 #12200 #14731

### CODE REVIEW

**Comment by ryantxu** [comment]
In #15310, @simPod and @buroa reported they did not see lazy loading.

This patch is up-to-date with master and I tested with this dashboard:
https://gist.github.com/ryantxu/a832f416c59f02127dca82cdcdd1b7a7

On OSX with chrome, safari, and firefox, I see these lines in the console as I scroll down:
image

**Comment by ryantxu** [comment]
Looks like circleci is failing on prettier issue with: public/app/plugins/datasource/graphite/FunctionEditor.tsx, I'll ignore that for now

**Comment by buroa** [comment]
Trying now.

**Comment by buroa** [comment]
Ok -- I see it working on my end. But with some bugs:

If I load a big dashboard (we automatically generate dashboards on the fly in my org) and scroll relatively fast, it just has a bunch of tiles (graphs) that are blank... They are supposed to be there, but it looks like they are moved further down the screen.

**Comment by ryantxu** [comment]
@buroa -- can you try with this patch?  I found issues with the approach of trying to calculate location with just the gridPos.  In particular it is not accurate until after it has been returned by the layout engine. 

The updates use the raw element to calculate position and should be accurate.

**Comment by simPod** [comment]
Now it seems lazyloading is being triggered properly 👍

File: public/app/features/dashboard/dashgrid/DashboardGrid.tsx [hunk 1]
```diff
@@ -187,18 +190,81 @@ export class DashboardGrid extends PureComponent<Props> {
...
+    const elem = this.panelRef[panel.id.toString()];
+    if (!elem) {
+      // NOTE the gridPos is also not valid until after the first render
+      // since it is passed to the layout engine and made to be valid
+      // for example, you can have Y=0 for everything and it will stack them
+      // down vertically in the second call
+      return false;
+    }
+
+    // TODO? Is there a better way to get this?
```
Comment (torkelo) [inline public/app/features/dashboard/dashgrid/DashboardGrid.tsx]: Sadly there is no easy way to get panel position and dimensions from the grid component, https://github.com/STRML/react-grid-layout/issues/506 . which is sad as it has all that info

Comment (torkelo) [reply public/app/features/dashboard/dashgrid/DashboardGrid.tsx]: hm.. Maybe we should open a PR to pass it on to children, if that is possible

**Review by torkelo — COMMENTED** [review]
Overall this looks like a really solid approach. With some more testing (manual) we can get this into v6.1

File: public/app/features/dashboard/dashgrid/DashboardGrid.tsx (hunk 1 above)
Comment (ryantxu) [reply public/app/features/dashboard/dashgrid/DashboardGrid.tsx]: I don't see any clear way to add this in  STRML/react-grid-layout.  The position is calculated in render() and manipulates raw dom style.  We could modify the `onLayoutChange` callback to include Position, but that starts getting a little ugly and would need to recalculate the position anyway.

**Comment by simPod** [comment]
I works pretty well but still... :D 

When you have eg. 30 panels, each loads 5s, 4 panels fits on one page. That means there is 8 pages in total. You have to scroll like 5-6 times (pages) with this mechanism to load them all. And after each scroll you have to wait for panels to load.

It should be possible to disable lazyloading as it's great feature but only for dashboards with panels that loads < 0.5s. When the load time is higher, it gets annoying.

**Comment by ryantxu** [comment]
@torkelo -- what do you think the next step is with this?

**Comment by torkelo** [comment]
What to find time to test & review this, adding it to next 6.2 release

**Comment by ryantxu** [comment]
Not that we need more to think about... but a nice side effect is that we would have all the info required to make a nice scrollspy style TOC like:
https://london.my-netdata.io/default.html#menu_apps_submenu_swap;theme=slate
image


https://medium.com/front-end-weekly/anatomy-of-a-scrollspy-component-with-react-and-typescript-1-2-c04f9d5c9bad

**Pull request merged**